src/main/java/kafdrop/service/KafkaHighLevelConsumer.java

```diff
@@ -196,20 +188,141 @@ synchronized List<ConsumerRecord<String, String>> getLatestRecords(String topic,
-        rec.timestampType(),
-        rec.serializedKeySize(),
-        rec.serializedValueSize(),
-        deserialize(deserializers.getKeyDeserializer(), rec.key()),
-        deserialize(deserializers.getValueDeserializer(), rec.value()),
-        rec.headers(),
-        rec.leaderEpoch()))
-      .collect(Collectors.toList());
+      .map(rec -> createConsumerRecord(rec, deserializers))
+      .toList();
+  }
+
+  /**
+   * Searches records from all partitions of a given topic containing a given text.
+   *
+   * @param topic          The topic
+   * @param searchString   Searched text.
+   * @param maximumCount   The maximum number of results to return
+   * @param startTimestamp The message timestamp to search from
+   * @param deserializers  Key and Value deserializers
+   * @return A list of consumer records for a given topic.
+   */
+  synchronized SearchResults searchRecords(String topic,
+                                           String searchString,
+                                           Integer maximumCount,
+                                           Date startTimestamp,
+                                           Deserializers deserializers) {
+    initializeClient();
+    final List<TopicPartition> partitions = determinePartitionsForTopic(topic);
+    kafkaConsumer.assign(partitions);
+    seekToTimestamp(partitions, startTimestamp);
+
+    return searchRecords(searchString, maximumCount, deserializers);
+  }
+
+  private void seekToTimestamp(List<TopicPartition> partitions, Date startTimestamp) {
+    kafkaConsumer.offsetsForTimes(partitions.stream()
+        .collect(Collectors.toMap(tp -> tp, tp -> startTimestamp.getTime())))
+      .forEach(this::seekToOffset);
   }
 
+  @NotNull
+  private SearchResults searchRecords(String searchString, Integer maximumCount, Deserializers deserializers) {
+    final List<ConsumerRecord<String, String>> foundRecords = new ArrayList<>();
+    boolean moreRecords;
+    var loweredSearchString = searchString.toLowerCase();
+    var scanStatus = new ScanStatus(Long.MAX_VALUE, 0);
+
+    do {
+      final var polled = kafkaConsumer.poll(Duration.ofMillis(SEARCH_POLL_TIMEOUT_MS));
+      scanStatus = searchPolledRecords(polled, deserializers, loweredSearchString, foundRecords, scanStatus);
+
+      moreRecords = !polled.isEmpty();
+    } while (foundRecords.size() < maximumCount
+      && moreRecords
+      && scanStatus.scannedCount() < SEARCH_MAX_MESSAGES_TO_SCAN);
+
+
+    CompletionReason completionReason;
+    if (!moreRecords) {
+      completionReason = CompletionReason.NO_MORE_MESSAGES_IN_TOPIC;
+    } else if (foundRecords.size() >= maximumCount) {
+      completionReason = CompletionReason.FOUND_REQUESTED_NUMBER_OF_RESULTS;
+    } else if (scanStatus.scannedCount() >= SEARCH_MAX_MESSAGES_TO_SCAN) {
+      completionReason = CompletionReason.EXCEEDED_MAX_SCAN_COUNT;
+    } else {
+      // TODO: This situation cannot occur. There is no exit criterion on end of timespan
```

Comment:
@nathanscottdaniels See this TODO. At the moment, there is no way the search ends on reaching the end of the timespan. If it would, it would result in the below message to the user:

```
Search reached the end of the specified time span before finding requested number of results.  Scanned  %d messages.
```

How do you want to resolve this? Add the search criterion or drop the end state? The latter, I can do for  you. For the former, I'd have to depend on you.

Reply:
@nathanscottdaniels Done